File: subprojects/docs/src/samples/userguide/plugins/readOnlyNestedProperty/java/buildSrc/src/main/java/UrlProcess.java
import org.gradle.api.DefaultTask;
import org.gradle.api.tasks.Nested;
import org.gradle.api.tasks.TaskAction;

// tag::url-process[]
public abstract class UrlProcess extends DefaultTask {
    // Use an abstract getter method annotated with @Nested
    @Nested
    abstract HostAndPath getHostAndPath();

    @TaskAction
    void run() {
        // Use the `hostAndPath` property
        System.out.println("Downloading https://" + getHostAndPath().getHostName().get() + "/" + getHostAndPath().getPath().get());
    }
}
// end::url-process[]

File: subprojects/docs/src/samples/userguide/plugins/readOnlyNestedProperty/kotlin/buildSrc/src/main/kotlin/UrlProcess.kt
import org.gradle.api.DefaultTask
import org.gradle.api.provider.Property
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.Nested
import org.gradle.api.tasks.TaskAction

// tag::url-process[]
abstract class UrlProcess : DefaultTask() {
    // Use an abstract getter method annotated with @Nested
    @get:Nested
    abstract val hostAndPath: HostAndPath

    @TaskAction
    fun run() {
        // Use the `hostAndPath` property
        println("Downloading https://${hostAndPath.hostName.get()}/${hostAndPath.path.get()}")
    }
}

interface HostAndPath {
    @get:Input
    val hostName: Property<String>
    @get:Input
    val path: Property<String>
}
// end::url-process[]
